### PR TITLE
Updating router lookahead

### PR DESCRIPTION
**Pull request by amin1377**
In this pull request, we have optimized the memory footprint and reduced the build time of the router lookahead by adopting a strategy that involves fewer samples during the construction process. The previous implementation stored the cost to reach each dx and dy, which has now been replaced with a more efficient approach that samples a reduced number of points on the chip.

This got bigger:
- Fixed a bug where the router lookahead Djikstra flood had "holes" (zero cost estimates) in a few places due to not finding any IPIN at the relevant (dx,dy).  Now we fill in those 0 entries with the average over a 3x3 window of nearby (dx,dy) values.
- Added a new place_delay_model (simple place delay model) which gets the minimum delay for each (dx,dy) by post-processing the router lookahead data. This simplifies code paths and greatly reduces place_delay_matrix compute time.

### CODE REVIEW

**Comment by amin1377** [comment]
Titan QoR (using map router lookahead)
Screenshot from 2023-12-15 11-05-01

**Comment by amin1377** [comment]
largest VTR QoR:
Screenshot from 2023-12-18 10-04-42

**Comment by amin1377** [comment]
QoR on large VTR benchmark:
Screenshot from 2024-01-12 11-54-11

**Comment by amin1377** [comment]
QoR on Titan benchmark:
Screenshot from 2024-01-12 13-43-32

**Comment by vaughnbetz** [comment]
One QoR failure (too good):
19:29:43 | k6_frac_N10_frac_chain_mem32K_40nm.xml/LU8PEEng.v/common min_chan_width_route_time relative value 0.09854500000000001 outside of range [0.1,15.0], above absolute threshold 2.0 and not equal to golden value: 4000.0
19:29:43 | regression_tests/vtr_reg_nightly_test3_odin/vtr_reg_qor...[Pass]

Just needs an updated golden.

As a follow-up, you should make a PR to enable the new placer delay matrix ASAP as the results are good and we don't want anyone to break it.  That PR should include VTR, Titan and Koios QoR results.

File: vpr/src/route/router_lookahead_cost_map.cpp [hunk 1]
```diff
@@ -342,11 +342,21 @@ std::vector<std::pair<int, int>> CostMap::list_empty() const {
 }
 
 static void assign_min_entry(util::Cost_Entry* dst, const util::Cost_Entry& src) {
```
[the remark this answers is not in the document]
Comment (vaughnbetz) [reply vpr/src/route/router_lookahead_cost_map.cpp:344]: Why do we pass a pointer to dst (now current_entry) and a ref to src?  If not much work, better to make both references.

File: vpr/src/route/router_lookahead_extended_map.h [hunk 2]
```diff
@@ -103,6 +103,10 @@ class ExtendedMapLookahead : public RouterLookahead {
     void write_intra_cluster(const std::string& /*file*/) const override {
         VPR_THROW(VPR_ERROR_ROUTE, "ExtendedMapLookahead::write_intra_cluster unimplemented");
     }
+
+    float get_opin_distance_min_delay(int /*physical_tile_idx*/, int /*from_layer*/, int /*to_layer*/, int /*dx*/, int /*dy*/) const override {
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_extended_map.h:107]: Comment this (not used for the extended_map_lookahead)

File: vpr/src/route/router_lookahead_map.cpp [hunk 3]
```diff
@@ -1021,358 +638,48 @@ static Cost_Entry get_nearby_cost_entry(int from_layer_num, int x, int y, int to
...
-     * when a rr node appears on different sides
-     * However, current test show that the simple strategy provides
-     * a good trade-off between runtime and quality of results
-     */
-    e_side rr_side = NUM_SIDES;
-    for (const e_side& candidate_side : SIDES) {
-        if (rr_graph.is_node_on_specific_side(rr, candidate_side)) {
-            rr_side = candidate_side;
-            break;
+static util::Cost_Entry get_nearby_cost_entry_average_neighbour(int from_layer_num,
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map.cpp:641]: Make sure this is declared at the top of the file and has a comment.

File: vpr/src/route/router_lookahead_map.cpp [hunk 4]
```diff
@@ -1021,358 +638,48 @@ static Cost_Entry get_nearby_cost_entry(int from_layer_num, int x, int y, int to
...
-            for (size_t y = 0; y < router_cost_map.dim_size(2); y++) {
-                VTR_LOG("(%zu,%zu,%zu):\n", x, y, to_layer_num);
 
-                for (size_t i = 0; i < router_cost_map[to_layer_num][x][y].cost_vector.size(); ++i) {
-                    Cost_Entry entry = router_cost_map[to_layer_num][x][y].cost_vector[i];
-                    VTR_LOG("  %d: delay=%10.3g cong=%10.3g\n", i, entry.delay, entry.congestion);
-                }
-            }
-        }
+    if (neighbour_num >= 3) {
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map.cpp:674]: Needs a comment. Consider if 3 should be a defined constant so it has a name.

File: vpr/src/route/router_lookahead_map.cpp [hunk 5]
```diff
@@ -1483,80 +791,77 @@ static void min_global_cost_map(vtr::NdMatrix<util::Cost_Entry, 3>& internal_opi
...
+    int num_layers = g_vpr_ctx.device().grid.get_num_layers();
+    int width = (int)g_vpr_ctx.device().grid.width();
+    int height = (int)g_vpr_ctx.device().grid.height();
+    distance_min_cost.resize({static_cast<unsigned long>(num_tile_types),
+                              static_cast<unsigned long>(num_layers),
+                              static_cast<unsigned long>(num_layers),
+                              static_cast<unsigned long>(width),
+                              static_cast<unsigned long>(height)});
+
+    for (int tile_type_idx = 0; tile_type_idx < num_tile_types; tile_type_idx++) {
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map.cpp:812]: Probably should comment the overall strategy here.

File: vpr/src/route/router_lookahead_map.h [hunk 6]
```diff
@@ -11,46 +11,31 @@ class MapLookahead : public RouterLookahead {
     explicit MapLookahead(const t_det_routing_arch& det_routing_arch, bool is_flat);
 
   private:
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map.h:13]: Take a crack at Doxygen comments for all functions etc. here if you can .... use in the new VPR internals for router lookahead tab.
Also make sure that there is a high level description of the router lookahead (and the various versions) in that new Doxygen documentation.

File: vpr/src/route/router_lookahead_map_utils.h [hunk 7]
```diff
@@ -25,8 +25,49 @@
 #include "rr_node.h"
 #include "rr_graph_view.h"
 
+/* we will profile delay/congestion using this many tracks for each wire type */
+#define MAX_TRACK_OFFSET 16
+
 namespace util {
 
+class Cost_Entry;
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map_utils.h:33]: Doxygen comments for these could be useful too.

Comment (amin1377) [reply vpr/src/route/router_lookahead_map_utils.h:33]: Will do doxygen commenting on lookahead utils in another PR.

**Pull request merged**